tests/Functional/FileSyncer/FileSyncerTestCase.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Tests\Functional\FileSyncer;
+
+use PhpTuf\ComposerStager\Infrastructure\FileSyncer\FileSyncerInterface;
+use PhpTuf\ComposerStager\Tests\Functional\TestCase;
+
+abstract class FileSyncerTestCase extends TestCase
+{
+    public static function tearDownAfterClass(): void
+    {
+        self::removeTestEnvironment();
+    }
+
+    abstract protected function createSut(): FileSyncerInterface;
+
+    /**
+     * @dataProvider providerDirectories
+     */
+    public function testSync($activeDir, $stagingDir): void
+    {
+        self::removeTestEnvironment();
+        self::createTestEnvironment($activeDir);
+
+        self::createFiles($activeDir, [
+            'no_changes_ever.txt',
+            'exclude_dir/to_test_recursion.txt',
+            'exclude_dir/and_change_file_in_active_dir_after_syncing_to_staging_dir.txt',
+            'exclude_file.txt',
+            'change_in_staging_dir_before_syncing_back_to_active_dir.txt',
+            'delete_from_staging_dir_before_syncing_back_to_active_dir.txt',
+            'very/deeply/nested/file/that/is/never/changed/in/either/the/active/directory/or/the/staging/directory.txt',
```

Comment:
I'm not being facetious when I tell you that I adore these file names. They are super clear and straightforward and say exactly what they're meant to test.

Reply:
Thanks! 😄